evm-contracts/src/v0.6/dev/Flags.sol

```diff
@@ -0,0 +1,122 @@
+pragma solidity 0.6.6;
+
+
+import "../Owned.sol";
+import "../dev/SimpleAccessControl.sol";
+
+
+/**
+ * @title The Flags contract
+ * @notice Allows flags to signal to any reader on the access control list.
+ * The owner can set flags, or designate other addresses to set flags. The
+ * owner must turn the flags off, other setters cannot.
+ */
+contract Flags is Owned, SimpleAccessControl {
+
+  mapping(address => bool) private flags;
+  mapping(address => bool) private setters;
+
+  event FlagOn(
+    address indexed subject
+  );
+  event FlagOff(
+    address indexed subject
+  );
+  event SetterEnabled(
+    address indexed setter
+  );
+  event SetterDisabled(
+    address indexed setter
+  );
+
+  /**
+   * @notice read the warning flag status of a contract address.
+   * @param subject The contract address being checked for a warning.
+   */
+  function getFlag(address subject)
+    public
+    view
+    checkAccess()
+    returns (bool)
+  {
+    return flags[subject];
+  }
+
+  /**
+   * @notice allows owner to enable the warning flags for mulitple addresses.
+   * @param subjects List of the contract addresses whose flag is being raised
+   */
+  function setFlagsOn(address[] calldata subjects)
+    external
+    returns (bool)
+  {
+    require(msg.sender == owner || setters[msg.sender], "Only callable by enabled setters");
+
+    for (uint256 i = 0; i < subjects.length; i++) {
+      address subject = subjects[i];
+
+      if (!flags[subject]) {
+        flags[subject] = true;
+        emit FlagOn(subject);
+      }
+    }
+  }
+
+  /**
+   * @notice allows owner to disable the warning flags for mulitple addresses.
+   * @param subjects List of the contract addresses whose flag is being lowered
+   */
+  function setFlagsOff(address[] calldata subjects)
+    external
+    onlyOwner()
+    returns (bool)
+  {
+    for (uint256 i = 0; i < subjects.length; i++) {
+      address subject = subjects[i];
+
+      if (flags[subject]) {
+        flags[subject] = false;
```

Comment:
We could just use `delete flags[subject];` here, I am not too sure we save any gas checking if the value is set. We should test for gas usage in both cases.

It also depends on the API we want to expose. If we look at `SimpleAccessControl` contract and `removeAccess` function, for example, it emits an event regardless if the user had access or not. The event is emitted every time the function is called, which is different behavior than this contract.

Reply:
Re: events, I think our current policy is to only emit an event when there's a state change.  There are some contracts that don't abide by this rule right now, but I imagine we'll fix them next time they're updated.

Reply:
The `if` condition is more to prevent us from emitting the log event. We want to only emit it when the state changes, and otherwise just treat it as a no-op. You're right that we don't follow that pattern for SAC, I think we may actually want to change the pattern for SAC. I'll look into this.

Reply:
I fixed the SimpleReadAccessController to only emit logs when the state actually changes.

Reply:
We vendored solidity our dependencies to prevent them from changing without our knowledge. I think keeping them all locally also works better with some solidity tooling, since not everything necessarily knows how to look in the `node_modules` directory to resolve certain paths.